lib/IRGen/LoadableByAddress.cpp

```diff
@@ -0,0 +1,1538 @@
+        auto currCase = instr->getCase(i);
+        auto *currBB = currCase.second;
+        SILBuilder argBuilder(currBB->begin());
+        assert(currBB->getNumArguments() <= 1 && "Unhandled BB Type");
+        EnumElementDecl *decl = currCase.first;
+        for (SILArgument *arg : currBB->getArguments()) {
+          SILType storageType = arg->getType();
+          SILType newSILType = getNewSILType(storageType, pass.Mod);
+          if (storageType == newSILType) {
+            newSILType = newSILType.getAddressType();
+          }
+
+          auto *newArg = argBuilder.createUncheckedTakeEnumDataAddr(
+              instr->getLoc(), instr->getOperand(), decl,
+              newSILType.getAddressType());
+          arg->replaceAllUsesWith(newArg);
+          currBB->eraseArgument(0);
+
+          // Load the enum addr then see if we can get rid of the load:
+          LoadInst *loadArg = nullptr;
+          if (pass.F->hasUnqualifiedOwnership()) {
+            loadArg = argBuilder.createLoad(
+                newArg->getLoc(), newArg, LoadOwnershipQualifier::Unqualified);
+          } else {
+            loadArg = argBuilder.createLoad(newArg->getLoc(), newArg,
+                                            LoadOwnershipQualifier::Take);
+          }
+          newArg->replaceAllUsesWith(loadArg);
+          loadArg->setOperand(newArg);
+          if (allUsesAreReplaceable(loadArg, true, pass.Mod)) {
+            allocator.replaceLoadWithCopyAddr(loadArg, true);
+          } else {
+            allocator.replaceLoadWithCopyAddrForApplies(loadArg);
+          }
+        }
+        caseBBs.push_back(std::make_pair(decl, currBB));
+      }
+      SILBasicBlock *defaultBB =
+          instr->hasDefault() ? instr->getDefaultBB() : nullptr;
+      auto *newInstr = enumBuilder.createSwitchEnumAddr(
+          instr->getLoc(), instr->getOperand(), defaultBB, caseBBs);
+      instr->replaceAllUsesWith(newInstr);
+      instr->getParent()->erase(instr);
+    }
+    while (!pass.structExtractInstsToMod.empty()) {
+      auto *instr = pass.structExtractInstsToMod.pop_back_val();
+      bool updateResultTy = pass.resultTyInstsToMod.count(instr) != 0;
+      if (updateResultTy) {
+        pass.resultTyInstsToMod.remove(instr);
+      }
+      SILBuilder structBuilder(instr);
+      auto *newInstr = structBuilder.createStructElementAddr(
+          instr->getLoc(), instr->getOperand(), instr->getField(),
+          instr->getType().getAddressType());
+      // Load the struct element then see if we can get rid of the load:
+      LoadInst *loadArg = nullptr;
+      if (pass.F->hasUnqualifiedOwnership()) {
+        loadArg = structBuilder.createLoad(newInstr->getLoc(), newInstr,
+                                           LoadOwnershipQualifier::Unqualified);
+      } else {
+        loadArg = structBuilder.createLoad(newInstr->getLoc(), newInstr,
+                                           LoadOwnershipQualifier::Take);
```

Comment:
Hmmm @gottesmm , I feel you are a better authority on this subject, is my `struct_extract` assumption correct?